epitator/infection_annotator.py

```diff
@@ -0,0 +1,314 @@
+            warning("Source span does not contain a noun chunk.")
+            nugget = doc.tiers["spacy.tokens"]
+            nc_tokens = []
+        else:
+            if len(ncs) > 1:
+                warning("Source span contains more than one noun chunk.")
+            nugget = ncs[0]
+
+        nc_tokens = [token for token in nugget.span]
+        tokens = nc_tokens
+        metadata = merge_dicts([
+            generate_attributes(tokens),
+            generate_counts(tokens, strict_only=False)
+        ], unique=True)
+        sources = []
+
+        if meets_inclusion_criteria(metadata):
+            sources = ["noun_chunk"]
+
+        # If the noun chunk is the subject of the root verb, we check the
+        # ancestors for metadata lemmas too.
+            if "nsubj" in [t.dep_ for t in nc_tokens]:
+                ancestors = [a for a in nugget.span.root.ancestors]
+                ancestor_metadata = merge_dicts([
+                    generate_attributes(ancestors),
+                    generate_counts(ancestors)
+                ], unique=True)
+                if meets_inclusion_criteria(ancestor_metadata):
+                    tokens.extend(ancestors)
+                    metadata = merge_dicts([metadata, ancestor_metadata],
+                                           unique=True)
+                    sources.append("ancestors")
+
+        # If the noun chunk's metadata indicates that it refers to a person,
+        # we check the disjoint subtree.
+        if "person" in metadata["attributes"]:
+            sources = ["noun_chunk"]
+            disjoint_subtree = [w for w in nugget.span.subtree if w.i not in [w.i for w in nugget.span]]
+            subtree_metadata = merge_dicts([
+                generate_attributes(disjoint_subtree),
+                generate_counts(disjoint_subtree)
+            ], unique=True)
+
+            if meets_inclusion_criteria(subtree_metadata):
+                # TODO: Consider iterating through until a triggering word is
+                # found.
+                tokens.extend(disjoint_subtree)
+                metadata = merge_dicts([metadata, subtree_metadata],
+                                       unique=True)
+                sources.append("subtree")
+
+        if include_sources:
+            metadata["sources"] = sources
+
+        # Generate counts from the noun chunk.
+        # counts = generate_counts(nc_tokens)
+        # metadata = merge_dicts([metadata, counts])
+
+        # Is "count" at most one value?
+        if "count" in metadata.values() and isinstance(metadata["count"], list):
+            raise TypeError("Metadata includes multiple count values.")
+
+        self.start = min([w.idx for w in tokens])
+        self.end = max([w.idx + len(w) for w in tokens])
+        self.doc = doc
+        self.metadata = metadata
+        self.label = self.text
+
+
+class InfectionAnnotator(Annotator):
+    def __init__(self, inclusion_filter=["infection", "death", "hospitalization"]):
```

Comment:
I went back and forth on this. I guess, it depends on how you frame what the annotator does.

If its goal is to produce a tier of annotations which correspond roughly to mentions of infection events in the document, then it makes sense to have the filter occur in the Annotate method. If the goal is to add annotations that can be filtered to produce a tier of infection events, then no filtering should occur.

I tend to lean towards the former approach — at least as the default behavior. The annotator is already making a lot of decisions in the Annotate method about a lot of things, and it would seem weird to defer one of those till _after_ the tier is created. Returning everything just adds in junky annotations that are likely irrelevant. For instance, it'd contain a bunch of spans with no properties, and might include a lot of person lemmas that have nothing to do with infections (which, if we wanted to generate, should probably have their own `PersonAnnotator()` tier). However, I included this option kinda as a debug feature. Do you think in this case it might be better to just have a single flag, something like "return_all_candidate_spans=False"?

Reply:
Yes, I think it makes the intended use of the parameter clearer.